test/externalTests/runners/foundry.py

```diff
@@ -0,0 +1,183 @@
+#!/usr/bin/env python3
+
+# ------------------------------------------------------------------------------
+# This file is part of solidity.
+#
+# solidity is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# solidity is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with solidity.  If not, see <http://www.gnu.org/licenses/>
+#
+# (c) 2023 solidity contributors.
+# ------------------------------------------------------------------------------
+
+import os
+import re
+import subprocess
+from pathlib import Path
+from shutil import which
+from string import Template
+from typing import Tuple
+from textwrap import dedent
+
+from exttest.common import AVAILABLE_PRESETS
+from exttest.common import settings_from_preset, get_solc_short_version
+from exttest.common import TestConfig, TestRunner
+
+
+def run_forge_command(command: str, env: dict = None):
+    subprocess.run(
+        command.split(),
+        env=env if env is not None else os.environ.copy(),
+        check=True
+    )
+
+
+class FoundryRunner(TestRunner):
+    """Configure and run Foundry-based projects"""
```

Comment:
It seems to me that the division of responsibilities between `TestRunner` and  `FoundryRunner` is quite unbalanced. `FoundryRunner` does pretty much everything and `TestRunner` is almost empty. I'd expect it to be the opposite, with the general test flow being shared for between all tests and `FoundryRunner` only adding some small deviations to account for how Foundry works.

After all you have all tests share the same `TestConfig` structure and accept the `_fn` callbacks (which TBH I'm not a fan of and I'd rather get rid of them) which abstract the setup/compilation/testing. What's the point of sharing these things if every test runner does things its own way?

Reply:
It was part of my initial design, but now that we switch to free functions to run the tests and setup everything, we may be able to get rid of it. The only place that actually depends on a common interface between any runner (e.g. Hardhat, Foundry and Truffle) is the `run_test` function. But I'm not really sure how to get rid of it without repeating some code on each runner, since the logic to run the tests is the same for all runners.

Reply:
> The callback functions are because some of the external tests, like uniswap and Gnosis Protocol, require a bunch of specific settings/workarounds

Yeah, I know. I just mean that doing this via callbacks does not feel like the best design. A more natural design would be to have these as virtual methods that you can override when needed. This would also be more flexible since you could invoke base method if needed and just add functionality (rather than completely replace it).

> The idea was that we could have one test framework that supports multiple types of runners which could still flexible enough to workarounds/fixes

That's fine but what I'm saying is that currently it seems like the framework is empty and almost all code is written as one big workaround :) There should be more shared code, even at this stage.

Reply:
> Yeah, I know. I just mean that doing this via callbacks does not feel like the best design. A more natural design would be to have these as virtual methods that you can override when needed. This would also be more flexible since you could invoke base method if needed and just add functionality (rather than completely replace it).

But they should be a virtual method of what? Because there is no similarity on how a hardhat-based project and a foundry-based project are configured for example. The only thing that they share is that they both have a configuration file and some script/command to build/test it. This was what the `TestRunner` was supposed to abstract away and in the original design, it was like you said, most of the free functions that we have now were methods of the `TestRunner`. We can revert that and just make `PRBMath` a class again that implements the `FoundryRunner` methods.

> That's fine but what I'm saying is that currently it seems like the framework is empty and almost all code is written as one big workaround :) There should be more shared code, even at this stage.

haha the only shared part between runners is the logic of the `run_test` method and the fact that they have a runner and compiler configuration. Each external test, on the other hand, share a lot of a runner (i.e. a hardhat-based project will only have some minor changes from other in the way that it is built, run or some workaround that we may need to do). This is currently done using callbacks, but we could have a class for each external test that extends a runner.

Reply:
After going over the PR again, I think I see what you actually intended with `TestRunner`. I assumed it was meant to contain both the framework-agnostic logic and some extension points to be filled in by framework-specific derived classes. But looks like it consists only of those extension points and the agnostic logic is all in standalone functions?

> But they should be a virtual method of what? Because there is no similarity on how a hardhat-based project and a foundry-based project are configured for example.

Of `TestRunner`. Then `FoundryRunner` would call those methods at the right points like it now does with callbacks. The only thing that would change would be how these methods are provided. I.e. user would derive its own customized runner from `FoundryRunner` and there provide these methods. It's basically the same mechanism you're already using with `TestRunner`, just those methods would be implemented by a particular test, not by the framework-specific runner.

Now that I think of it though, these callbacks are really something we need only for JS-based frameworks. For Foundry they'll pretty much always go empty. So I think that instead of being `TestRunner`'s virtual methods, they could be `HardhatRunner`'s methods, to be added only in later PR. We could omit them from this PR, and for Hardhat maybe we'll even find a way to make it more generic.

Reply:
> But looks like it consists only of those extension points and the agnostic logic is all in standalone functions?

Yes. Before it had a bit of the framework-agnostic logic as well, but this changed on the course of this PR.

> So I think that instead of being `TestRunner`'s virtual methods, they could be `HardhatRunner`'s methods, to be added only in later PR. We could omit them from this PR, and for Hardhat maybe we'll even find a way to make it more generic.

Right. I changed the current code to accommodate that in the future when we migrate the hardhat-based tests.